src/Storages/MergeTree/IMergeTreeDataPart.cpp

```diff
@@ -1325,6 +1338,91 @@ String IMergeTreeDataPart::getRelativePathForDetachedPart(const String & prefix)
     return "detached/" + getRelativePathForPrefix(prefix);
 }
 
+void IMergeTreeDataPart::addVirtualProjectionPart(const StorageMetadataPtr & metadata_snapshot)
+{
+    const auto & projections = metadata_snapshot->getProjections();
+    if (!projections.has(ProjectionDescription::VIRTUAL_PROJECTION_NAME))
+        return;
+
+    auto block = projections.get(ProjectionDescription::VIRTUAL_PROJECTION_NAME).sample_block;
+    if (!minmax_idx.initialized)
+    {
+        if (isEmpty())
+        {
+            auto part_virtual = std::make_shared<MergeTreeDataPartInMemory>(storage, block);
+            part_virtual->parent_part = this;
+            projection_parts.emplace(ProjectionDescription::VIRTUAL_PROJECTION_NAME, std::move(part_virtual));
+            return;
+        }
+        else
+            throw Exception(
+                "Cannot add virtual projection part when minmax_idx is not initialized yet. It's a bug", ErrorCodes::LOGICAL_ERROR);
+    }
+
+    auto minmax_count_columns = block.mutateColumns();
+    size_t minmax_idx_size = minmax_idx.hyperrectangle.size();
+    if (2 * minmax_idx_size != minmax_count_columns.size() - 1)
```

Comment:
Reading the code below, I think that we store `min` and `max` for each column, and also `count`.
However, looking reading this exception it is not clear.
Also, why don't we check `2 * minmax_idx_size + 1 != minmax_count_columns.size()`?